### PR TITLE
Add documentation on using Authenticator with Kops

### PR DESCRIPTION
**Pull request by rifelpet**
This addresses a comment in #31.

I have successfully added Authenticator with kops using hooks and a systemd service. I wasn't sure where you want this, If it should go in its own docs file I'm happy to move it there.

I'm also not positive that this is the "best" approach but it definitely works 😄  `fileAssets` or `execContainer` are also viable solutions as discussed in the issue.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -105,6 +105,45 @@ The token is valid for 15 minutes (the shortest value AWS permits) and can be re
...
+       before:
+         - kubelet.service
+       roles: [Master]
+       manifest: |
+         [Unit]
+         Description=Download Heptio AWS Authenticator configs from S3
+         [Service]
+         Type=oneshot
+         ExecStart=/bin/mkdir -p /srv/kubernetes/heptio-authenticator-aws
+         ExecStart=/usr/local/bin/aws s3 cp --recursive s3://KOPS_STATE_STORE/CLUSTER_NAME/addons/authenticator /srv/kubernetes/heptio-authenticator-aws/
```
Comment (phillipj) [inline README.md]: I had to add an explicit `--region` argument to this command.

**Comment by phillipj** [comment]
Really appreciated this example, helped me a lot to get the authenticator up-n-running in our clusters with kops 👍

**Comment by kdillane** [comment]
How did you boostrap the Authenticator DaemonSet and ConfigMap?  When creating a cluster using kops and using the authenticationTokenWebhookConfigFile kubeApiServer argument, there's no clear path to creating these resources.

**Comment by phillipj** [comment]
I started off with the [example.yaml](https://github.com/heptio/authenticator/blob/master/example.yaml) referenced in the [Run the server section](https://github.com/heptio/authenticator#2-run-the-server) of the `README.md`. Did some changes to the ConfigMap by trial-n-error.

Important to remember step 4 of the `README.md` changes proposed in this PR: 

> Update the Authenticator DaemonSet's state and output volumes to both use `/srv/kubernetes/heptio-authenticator-aws/` for their `hostPath`s.

**Comment by HagaiBarel** [comment]
Doesn't work while using coreos container linux. Turns out the ami doesn't have aws cli installed and therefore the `aws s3 cp` command fails

**Comment by kdillane** [comment]
Did you script the cluster creation end to end?  How did you create the ConfigMap and DaemonSet if the api server is protected by the Authenticator?

**Comment by rifelpet** [comment]
@kdillane other authentication methods still work, so having the certificate data in your kubeconfig works fine without any aws credentials. We deploy in a CI pipeline:
```
kops export kubecfg ${CLUSTER_NAME}
kubectl apply -f authenticator.yml
```

**Comment by kdillane** [comment]
Makes sense. Okay, I must have been running into a separate issue from authentication not working. I’ll dig in. Thanks for the feedback!

**Comment by sp-joseluis-ledesma** [comment]
I've used a slightly different approch, In the kops hook:
```
  - manifest: |
      Type=oneshot
      ExecStart=/bin/mkdir -p /etc/kubernetes/manifests /etc/heptio-authenticator-aws
      ExecStart=/bin/sh -c '/bin/echo <BASE64_CONFIG.YAML> |/usr/bin/base64 -d >/etc/heptio-authenticator-aws/config.yaml'
      ExecStart=/bin/sh -c '/bin/echo  <BASE64_POD.YAMLK>|/usr/bin/base64 -d >/etc/kubernetes/manifests/heptio-authenticator-aws.manifest'
      ExecStop=/bin/rm -f /etc/kubernetes/manifests/heptio-authenticator-aws.manifest
      RemainAfterExit=true
    name: heptio-authenticator-aws.service
    requires:
    - kubelet.service
    - protokube.service
    roles:
    - Master
```
where POD.YAML is a modified version of the original DaemonSet:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    scheduler.alpha.kubernetes.io/critical-pod: ""
  creationTimestamp: null
  labels:
    k8s-app: heptio-authenticator-aws
  name: heptio-authenticator-aws
  namespace: kube-system
spec:
  # run on the host network (don't depend on CNI)
  hostNetwork: true
  containers:
  - name: heptio-authenticator-aws
    image: gcr.io/heptio-images/authenticator:v0.1.0
    args:
    - server
    - --config=/etc/heptio-authenticator-aws/config.yaml
    - --state-dir=/var/heptio-authenticator-aws
    - --generate-kubeconfig=/srv/kubernetes/heptio-authenticator-aws/kubeconfig.yaml
    resources:
      requests:
        memory: 20Mi
        cpu: 10m
      limits:
        memory: 20Mi
        cpu: 100m

    volumeMounts:
    - name: config
      mountPath: /etc/heptio-authenticator-aws/
    - name: state
      mountPath: /var/heptio-authenticator-aws/
    - name: output
      mountPath: /srv/kubernetes/heptio-authenticator-aws/

  volumes:
  - name: config
    hostPath:
      path: /etc/heptio-authenticator-aws
  - name: output
    hostPath:
      path: /srv/kubernetes/heptio-authenticator-aws/
  - name: state
    hostPath:
      path: /var/heptio-authenticator-aws/
```

This would work for cases like using CoreOS AMI (@HagaiBarel), and allows using kops rolling-update cluster to update the config.yaml (without being necessary to force it)

File: README.md (hunk 1 above)
Comment (willejs) [reply README.md]: I am using 
```
ExecStart=/usr/bin/docker run --net=host --rm -v /srv/kubernetes/heptio-authenticator-aws:/srv/kubernetes/heptio-authenticator-aws quay.io/coreos/awscli@sha256:7b893bfb22ac582587798b011024f40871cd7424b9026595fd99c2b69492791d aws s3 cp --recursive {{.config_base}}/addons/authenticator /srv/kubernetes/heptio-authenticator-aws/
``` 
instead on core os

File: README.md [hunk 2]
```diff
@@ -105,6 +105,45 @@ The token is valid for 15 minutes (the shortest value AWS permits) and can be re
...
+   ```
+3. Apply the changes with `kops update cluster ${CLUSTER_NAME}`.
+   If the cluster already exists, roll the cluster with `kops rolling-update cluster ${CLUSTER_NAME}` in order to recreate the master nodes.
+4. Update the Authenticator DaemonSet's state and output volumes to both use `/srv/kubernetes/heptio-authenticator-aws/` for their `hostPath`s.
```
Comment (christopherhein) [inline README.md]: Probably want a 5th step or an addition to this that says to `apply` the `ConfigMap` and `DaemonSet` to be pedantic.

File: README.md (hunk 1 above)
Comment (christopherhein) [reply README.md]: I like the addition of the MasterMind templating in this, might be nice to pull that into the readme.

**Comment by christopherhein** [comment]
Another note for this as of right now is if you use the new `ExecCredential` configuration in your `kubecfg` it will get mangled when you create or delete a cluster using `kops`

**Comment by rifelpet** [comment]
@christopherhein Addressed both of your comments

**Comment by christopherhein** [comment]
@rifelpet Can you do one last thing, and add a note about using docker for non-default AMI deploys of `kops` like @willejs said…

```
ExecStart=/usr/bin/docker run --net=host --rm -v /srv/kubernetes/heptio-authenticator-aws:/srv/kubernetes/heptio-authenticator-aws quay.io/coreos/awscli@sha256:7b893bfb22ac582587798b011024f40871cd7424b9026595fd99c2b69492791d aws s3 cp --recursive {{.config_base}}/addons/authenticator /srv/kubernetes/heptio-authenticator-aws/
```

**Comment by christopherhein** [comment]
Thanks @rifelpet, @nckturner lgtm! :shipit:

**Comment by Raffo** [comment]
Been using the same suggestions, works for me... more or less. Unfortunately, I'm hitting #25 which makes the setup definitely not perfect, so a workaround should be mentioned in this guide or just integrated.

**Comment by christopherhein** [comment]
I've ran through this a number of more times. I would say we're good to go and if we need to make updates after when @Raffo gets a chance to retest that shouldn't be that big of a deal. Also we have this live which has been working well - https://aws.amazon.com/blogs/opensource/deploying-heptio-authenticator-kops/

**Comment by Raffo** [comment]
I agree, when I'll have the change to test it again I will definitely open an issue if I get another issue. Thanks for the great work!

**Comment by christopherhein** [comment]
@nckturner can we get this merged?

**Comment by StevenACoffman** [comment]
@rifelpet Do you have suggestions for how to secure the dashboard in kops using heptio-authenticator? `heptio-authenticator-aws  token -i $cluster -r $arn | jq -r '.status.token'` maybe?

**Comment by christopherhein** [comment]
@StevenACoffman check - https://github.com/heptio/authenticator/issues/84

**Comment by j00p34** [comment]
For anyone coming here from the AWS page, like me: The main readme https://github.com/kubernetes-sigs/aws-iam-authenticator has more up to date info.

**Pull request merged**